Add gelu op

xiaomengy opened the pull request:
Summary: Add gelu op

Differential Revision: D14814811

houseroad commented on caffe2/python/operator_test/activation_ops_test.py:
```diff
@@ -215,3 +217,27 @@ def leaky_relu_ref(X):
+            cdf = 0.5 * (1.0 + erf(X / np.sqrt(2.0)))
+            return (X * cdf,)
+
+        tol = 1e-3 if fast_gelu else 1e-4
```
do we want to try 1e-5 tolerance in non fast case.

BIT-silence replied:
The default value is 1e-4
https://github.com/pytorch/pytorch/blob/master/caffe2/python/hypothesis_test_util.py#L563

houseroad replied:
make sense. My concern is unnecessary. :-)

houseroad commented on caffe2/operators/gelu_op.cc:
```diff
@@ -0,0 +1,132 @@
+operator()(const int N, const T* X, T* Y, CPUContext* context) const {
+  ConstEigenVectorArrayMap<T> X_arr(X, N);
+  EigenVectorArrayMap<T> Y_arr(Y, N);
+  if (fast_gelu) {
```
add more comments? e.g., the equation of fast-gelu and gelu? maybe the link of the paper?

BIT-silence replied:
Done

houseroad commented on caffe2/operators/gelu_op.cc:
```diff
@@ -0,0 +1,132 @@
+OPERATOR_SCHEMA(Gelu)
+    .NumInputs(1)
+    .NumOutputs(1)
+    .Arg("fast_gelu", "If true, use fast approximation")
```
maybe some rough perf and accuracy difference in two modes?

houseroad commented on caffe2/operators/gelu_op.cc:
```diff
@@ -0,0 +1,132 @@
+  ConstEigenVectorArrayMap<T> X_arr(X, N);
+  EigenVectorArrayMap<T> dX_arr(dX, N);
+  if (fast_gelu) {
+    constexpr T kAlpha = gelu_utils::kSqrt2 / gelu_utils::kSqrtPi;
```
equations here?

houseroad reviewed (commented):
I think inplace case won't work with GELU. Shall we detect such case and error out if so?

I am also wondering what's the perf and accurary different between two modes (fast vs non-fast).

houseroad commented:
https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-cuda9.0-cudnn7-windows-build/40605/console

failing build is legit, please take a look

xiaomengy committed:
Add gelu op (#18992)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/18992

Add gelu op

Reviewed By: houseroad

Differential Revision: D14814811

fbshipit-source-id: f19e7e593a5394e19cb8b87cf04799610ec5c4ac
